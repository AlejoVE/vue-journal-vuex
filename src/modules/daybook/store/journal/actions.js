// export const myAction = async ({commit}) => {
// }

import journalApi from "@/api/journalApi"

export const loadEntries = async ({commit}) => {
    try {
        const {data} = await journalApi.get("/entries.json")

        if(!data) return commit("setEntries", [])

        const entries = []
        for (let id of Object.keys(data)) {
            entries.push({
                id,
                ...data[id]
            })
        }
        commit("setEntries", entries)
    } catch (error) {
        console.log('ERROR ACTION: loadEntries', error)
    }
}

export const createEntry = async ({commit}, entry) => {
    try {
        const {text, date, picture} = entry
        const dataToSave = {text, date, picture}
        const {data} = await journalApi.post(`/entries.json`, dataToSave)

        dataToSave.id = data.name

        commit("addEntry", dataToSave)

        return data.name
    } catch (error) {
        console.log('ERROR ACTION: createEntry', error)
    }
}

export const updateEntry = async ({commit}, entry) => {
    try {
        const {text, date, picture} = entry
        const dataToSave = {text, date, picture}
        await journalApi.put(`/entries/${entry.id}.json`, dataToSave)
        commit("updateEntry", {...entry})
    } catch (error) {
        console.log('ERROR ACTION: updateEntry', error)
    }
}

export const deleteEntry = async ({commit}, id) => {
    try {
        console.log('deleteEntry', id)
        await journalApi.delete(`/entries/${id}.json`)
        commit("deleteEntry", id)
    } catch (error) {
        console.log('ERROR ACTION: deleteEntry', error)
    }
}