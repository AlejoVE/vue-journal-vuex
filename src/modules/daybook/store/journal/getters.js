// export const myGetter =  (state) => {
    // return state.something
// }

export const getEntriesByTerm =  (state) => (searchTerm = '')=> {
   if(searchTerm.length === 0) return state.entries

   return state.entries.filter(entry => entry.text.toLowerCase().includes(searchTerm.toLowerCase()))
}

export const getEntryById =  (state) => (id) => {
   const entry = state.entries.find(entry => entry.id === id)

   if(!entry) return
    /* We do it this way because if we return  state.entries directly,
    it will be a reference to the state.entries array, and if we modify the entry,
     it will change the state.entries array as well.
    */
   return {...entry}
}
