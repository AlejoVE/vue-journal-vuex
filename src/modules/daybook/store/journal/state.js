export default () => ({
     isLoading: true,
     entries: [
         {
             id: new Date().getTime(),
             date: new Date().toDateString(),
             picture: null,
             text:  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin dictum, neque et pulvinar maximus, velit neque condimentum metus, a laoreet nunc enim vitae dui. Quisque maximus velit a vulputate egestas. Integer cursus mattis molestie. Mauris gravida sed purus vel viverra. Donec mauris nisi, egestas placerat quam dapibus, lacinia semper mi. Vivamus aliquet aliquet mauris,',
         },
         {
             id: new Date().getTime() + 1000,
             date: new Date().toDateString(),
             picture: null,
             text:  'eu dignissim velit pharetra eu. Aenean lacinia ut nulla in elementum. Ut eget placerat libero, egestas ullamcorper nisi. Curabitur porttitor diam vel congue semper. Vestibulum accumsan, nisl at congue volutpat, ante elit pellentesque dui, a finibus turpis libero vitae nunc. Vivamus tempus consequat',
         },
         {
             id: new Date().getTime() + 2000,
             date: new Date().toDateString(),
             picture: null,
             text:  'In porttitor ornare diam, at tincidunt neque rutrum vel. Sed sed porta enim, id varius nunc. Phasellus tristique, nunc sit amet malesuada porta, felis diam iaculis sem, ultricies rutrum magna urna id enim. Morbi mi lacus, elementum lobortis fringilla non, rutrum ac mauris.',
         },
     ],
})