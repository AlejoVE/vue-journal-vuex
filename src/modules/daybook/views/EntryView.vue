<template>
  <template v-if="entry">
    <div class="entry-title  d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{day}}</span>
        <span class="fs-3 mx-1">{{month}}</span>
        <span class="fs-4 mx-2 fw-light ">{{yearDay}}</span>
      </div>
      <div>
        <input type="file" @change="onSelectedImage" ref="imageSelector" v-show="false" accept="image/png, image/jpg, image/jpeg"/>
        <button v-if="entry.id" @click="onDeleteEntry" class="btn btn-danger">Delete <i class="fa fa-trash-alt"></i></button>
        <button @click="onSelectImage" class="btn btn-primary mx-2">Upload photo <i class="fa fa-upload"></i></button>
        <button class="btn btn-"></button>
      </div>
    </div>
    <hr />
    <div class="d-flex flex-column px-3 h-75">
      <textarea v-model="entry.text" placeholder="What happened today?"></textarea>
    </div>
    <Fab @on:click="saveEntry" icon="far fa-save"/>
    <img v-if='entry.picture && !localImage' :src="entry.picture" />
    <img v-if="localImage" :src="localImage" class="img-thumbnail" alt="entry-picture" />
  </template>
</template>

<script>
import { defineAsyncComponent } from 'vue'
import Swal from 'sweetalert2'
import { mapGetters, mapActions } from 'vuex'
import uploadImage from '../helpers/uploadImage'

import getDayMonthYear from '../helpers/getDayMonthYear'

export default {
  name: 'EntryView',
  props: {
    id: {
      type: String,
      required: true
    }
  },
  components: {
    Fab: defineAsyncComponent(() => import('../components/Fab.vue'))
  },

  data(){
    return {
      entry: null,
      localImage: null,
      file: null
    }
  },
  methods: {
    loadEntry() {

      let entry;

      if(this.id === 'new') {
        entry = {
          text: '',
          date: new Date().getTime()
        }
      } else {
        entry = this.getEntryById(this.id)
        if(!entry) return this.$router.push({name: 'no-entry'})
      }

      this.entry = entry
    },
    async saveEntry(){

      new Swal({
        title: 'Saving...',
        allowOutsideClick: false
      })

      Swal.showLoading()

      if(this.file) {
        const picture = await uploadImage(this.file)
        this.entry.picture = picture
      }

      if(this.entry.id){
        await this.updateEntry(this.entry)
      } else {
        const id = await this.createEntry(this.entry)
        this.$router.push({name: 'entry', params: { id }})
      }

      Swal.fire('Saved!', '', 'success')
    },
    async onDeleteEntry(){
      const { isConfirmed } =  await new Swal({
        title: 'Are you sure?',
        showDenyButton: true,
        denyButtonText: 'Cancel',
        confirmButtonText: "Yes, delete it!",
      })

      if(!isConfirmed) return
      Swal.showLoading()

      await this.deleteEntry(this.entry.id)
      this.$router.push({name: 'no-entry'})

      Swal.fire('Deleted!', '', 'success')
    },

    onSelectedImage(event){
      const file = event.target.files[0]
      if(!file) {
        this.localImage = null
        this.file = null
        return
      }
      this.file = file
      const fr = new FileReader()
      fr.onload = () => this.localImage = fr.result
      fr.readAsDataURL(file)
    },
    onSelectImage(){
      this.$refs.imageSelector.click()
    },
    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry'])
  },
  computed: {
    ...mapGetters('journal',['getEntryById']),
    day(){
      const {day} = getDayMonthYear(this.entry.date)
      return day
    },
    month(){
      const {month} = getDayMonthYear(this.entry.date)
      return month
    },
    yearDay(){
      const {yearDay} = getDayMonthYear(this.entry.date)
      return yearDay
    }
  },
  created(){
    this.loadEntry()
  },
  watch: {
    id(){
      this.loadEntry()
      this.file = null
      this.localImage = null
    }
  }
}
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>