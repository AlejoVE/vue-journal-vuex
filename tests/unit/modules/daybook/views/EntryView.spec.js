import { createStore } from "vuex"
import { shallowMount } from '@vue/test-utils'
import Swal from 'sweetalert2'
import journal from '@/modules/daybook/store/journal'
import journalState from '../../../mocks-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView'

const createVuexStore = (initialState) =>(
    createStore({
        modules: {
            journal: {
                ...journal,
                state: {...initialState}
            }
        }
    })
 )

 jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
 }))

describe('Tests in EntryView', () => {
    const store = createVuexStore(journalState)
    // To prevent actions to impact the database
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-MqpUtn-FdzZnSJMrIt3'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('should take out the user because the id does not exists', () => {
        const  wrapper = shallowMount(EntryView, {
            props: {
                id: 'sjnduiasnkld'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })

        expect(mockRouter.push).toBeCalledWith({name: 'no-entry'})
    })

    test('should show entry correctly', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toBeCalled()
    })

    test('should delete entry and exit', (done) => {
        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))
        wrapper.find('.btn-danger').trigger('click')

        expect(Swal.fire).toHaveBeenCalledWith({
            title: 'Are you sure?',
            showDenyButton: true,
            denyButtonText: 'Cancel',
            confirmButtonText: "Yes, delete it!",
        })

        setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-MqpUtn-FdzZnSJMrIt3')
            expect(mockRouter.push).toBeCalledWith({name: 'no-entry'})
            done()
        }, 1)

    })

})

