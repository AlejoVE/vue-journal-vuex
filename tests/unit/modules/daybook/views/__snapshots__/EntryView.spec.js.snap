// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in EntryView should show entry correctly 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">13</span><span class="fs-3 mx-1">Dec</span><span class="fs-4 mx-2 fw-light">2021, Mon</span></div>
  <div><input type="file" accept="image/png, image/jpg, image/jpeg" style="display: none;"><button class="btn btn-danger">Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary mx-2">Upload photo <i class="fa fa-upload"></i></button><button class="btn btn-"></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?"></textarea></div>
<fab-stub icon="far fa-save"></fab-stub>
<!--v-if-->
<!--v-if-->
`;
